src/server/shell/decoration/decoration_strategy.h

```diff
@@ -0,0 +1,199 @@
+};
+
+auto border_type_for(MirWindowType type, MirWindowState state) -> BorderType;
+
+struct ButtonInfo
+{
+    ButtonFunction function;
+    ButtonState state;
+    geometry::Rectangle rect;
+
+    auto operator==(ButtonInfo const& other) const -> bool
+    {
+        return function == other.function &&
+               state == other.state &&
+               rect == other.rect;
+    }
+};
+
+/// Describes the state of the interface (what buttons are pushed, etc)
+class InputState
+{
+public:
+    InputState(
+        std::vector<ButtonInfo> const& buttons,
+        std::vector<geometry::Rectangle> const& input_shape)
+        : buttons_{buttons},
+          input_shape_{input_shape}
+    {
+    }
+
+    auto buttons() const -> std::vector<ButtonInfo> const& { return buttons_; }
+    auto input_shape() const -> std::vector<geometry::Rectangle> const& { return input_shape_; }
+
+private:
+    InputState(InputState const&) = delete;
+    InputState& operator=(InputState const&) = delete;
+
+    std::vector<ButtonInfo> const buttons_;
+    std::vector<geometry::Rectangle> const input_shape_;
+};
+
+/// Information about the geometry and type of decorations for a given window
+/// Data is pulled from the surface on construction and immutable after that
+class WindowState
+{
+public:
+    WindowState(
+        std::shared_ptr<StaticGeometry> const& static_geometry,
+        std::shared_ptr<scene::Surface> const& window_surface,
+        float scale);
+
+    auto window_size() const -> geometry::Size;
+    auto border_type() const -> BorderType;
+    auto focused_state() const -> MirWindowFocusState;
+    auto window_name() const -> std::string;
+
+    auto titlebar_width() const -> geometry::Width;
+    auto titlebar_height() const -> geometry::Height;
+
+    auto side_border_width() const -> geometry::Width;
+    auto side_border_height() const -> geometry::Height;
+
+    auto bottom_border_width() const -> geometry::Width;
+    auto bottom_border_height() const -> geometry::Height;
+
+    auto titlebar_rect() const -> geometry::Rectangle;
+    auto left_border_rect() const -> geometry::Rectangle;
+    auto right_border_rect() const -> geometry::Rectangle;
+    auto bottom_border_rect() const -> geometry::Rectangle;
+
+    auto scale() const -> float;
+
+private:
+    WindowState(WindowState const&) = delete;
+    WindowState& operator=(WindowState const&) = delete;
+
+    std::shared_ptr<StaticGeometry> const static_geometry;
+    geometry::Size const window_size_;
+    BorderType const border_type_;
+    MirWindowFocusState const focus_state_;
+    std::string window_name_;
+    float scale_;
+};
+
+class RendererStrategy
+{
+public:
+    static auto alloc_pixels(geometry::Size size) -> std::unique_ptr<Pixel[]>;
+
+    RendererStrategy() = default;
+    virtual ~RendererStrategy() = default;
+
+    struct RenderedPixels
+    {
+        MirPixelFormat const format;
+        geometry::Size const size;
+        Pixel const* const pixels;
+    };
```

Comment:
Ah. I see I should add a `mrs::as_write_mappable_buffer()` to mirror `as_read_mappable_buffer()`; this should really just be `mg::Buffer`.

Reply:
I didn't want to force the decoration strategy to deal with the likes of `mrs::alloc_buffer_with_content()` which needs a `GraphicBufferAllocator`. Yes, it is all rendering, but this seemed like a tractable seam to insert in the design

Reply:
Yeah, `mrs::alloc_buffer_with_content()` is the wrong interface here; what we really want from the decorator is a `mg::Buffer`, as long as it's easy enough for them to render to it.

Then again, I've had broader thoughts about the whole requirements, which I'll need to coalesce and transmit in a different forum.